app/lib/features/chat/widgets/avatar_builder.dart

```diff
@@ -1,28 +1,31 @@
-import 'package:acter/common/providers/chat_providers.dart';
+import 'package:acter/common/providers/room_providers.dart';
 import 'package:acter_avatar/acter_avatar.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
 
 class AvatarBuilder extends ConsumerWidget {
-  final String userId;
+  late RoomMemberQuery query;
```

Comment:
Yeah, maybe you can help me with this. The problem is (originally) that I can't get providers to work with tuples. Now, to avoid that I wrap them into a custom type, so far, so good.

But of course I don't want the user of the widget to have to create/take care of that extra wrapper type, so I was creating it from  with `build` . But as it turns out (and cost me a day to debug) doing thst in create means the family provider considers it a new input and reevaluate in an endless loop.

So I have to plug it onto the widget instance to reuse the same instance. I don't quite understand the difference between the various instantion methods, bur this is the best I could find. If not like this, what is better to do? A stateful widget then?

Reply:
Generally [`StatelessWidget`](https://api.flutter.dev/flutter/widgets/StatelessWidget-class.html) is class that does not require mutable state. But in some use-case like in our case, if it required mutable data then we can just add `//ignore: must_be_immutable` above that class so that it can pass the Lint checking and user can come to know that this class require mutable data.

And I don't suggest `StatefulWidget` here in this case just because of link error of mutable params. it is ideally good to use `StatefulWidget` where we requires mutable data configuration along with the rebuilding UI states.

Suggestion:
- You can use `typedef` implementation for provider queries stuff. @gtalha07 recently implemented it on his [PR](https://github.com/acterglobal/a3/pull/1258).

Reply:
found a pretty solution thanks to @gtalha07 's work on the convo selection refactor